torch/distributions.py

```diff
@@ -0,0 +1,150 @@
+
+    Samples are binary (0 or 1). They take the value `1` with probability `p`
+    and `0` with probability `1 - p`.
+
+    Example::
+
+        >>> m = Bernoulli(torch.Tensor([0.3]))
+        >>> m.sample()  # 30% chance 1; 70% chance 0
+         0.0
+        [torch.FloatTensor of size 1]
+
+    Args:
+        probs (Tensor or Variable): the probabilty of sampling `1`
+    """
+    def __init__(self, probs):
+        self.probs = probs
+
+    def sample(self):
+        return torch.bernoulli(self.probs)
+
+    def log_prob(self, value):
+        # compute the log probabilities for 0 and 1
+        log_pmf = (torch.stack([1 - self.probs, self.probs])).log()
+
+        # evaluate using the values
+        return log_pmf.gather(0, value.unsqueeze(0).long()).squeeze(0)
+
+
+class Multinomial(Distribution):
+    r"""
+    Creates a multinomial distribution parameterized by `probs`.
+
+    Samples are integers from `0 ... K-1` where `K` is probs.size(-1).
+
+    If `probs` is 1D with length-`K`, each element is the relative probability
+    of sampling the class at that index.
+
+    If `probs` is 2D, it is treated as a batch of probability vectors.
+
+    See also: :func:`torch.multinomial`
+
+    Example::
+
+        >>> m = Multinomial(torch.Tensor([ 0.25, 0.25, 0.25, 0.25 ]))
+        >>> m.sample()  # equal probability of 0, 1, 2, 3
+         3
+        [torch.LongTensor of size 1]
+
+    Args:
+        probs (Tensor or Variable): event probabilities
+    """
+    def __init__(self, probs):
+        if probs.dim() != 1 and probs.dim() != 2:
+            # TODO: treat higher dimensions as part of the batch
+            raise ValueError("probs must be 1D or 2D")
+        self.probs = probs
+
+    def sample(self):
+        return torch.multinomial(self.probs, 1, True).squeeze(-1)
+
+    def log_prob(self, value):
+        p = self.probs / self.probs.sum(-1, keepdim=True)
+        if value.dim() == 1 and self.probs.dim() == 1:
+            # special handling until we have 0-dim tensor support
+            return p.gather(-1, value).log()
+
+        return p.gather(-1, value.unsqueeze(-1)).squeeze(-1).log()
+
+
+class Normal(Distribution):
+    r"""
+    Creates a normal (also called Gaussian) distribution parameterized by
+    `mean` and `std`.
+
+    Example::
+
+        >>> m = Normal(torch.Tensor([0.0]), torch.Tensor([1.0]))
+        >>> m.sample()  # normally distributed with mean=0 and stddev=1
+         0.1046
+        [torch.FloatTensor of size 1]
+
+    Args:
+        mean (Tensor or Variable): mean of the distribution
+        std (Tensor or Variable): standard deviation of the distribution
+    """
+    def __init__(self, mean, std):
+        if not torch.is_tensor(mean) and not isinstance(mean, Variable):
+            raise TypeError('mean must be a Tensor or Variable')
+        if not torch.is_tensor(std) and not isinstance(std, Variable):
+            raise TypeError('std must be a Tensor or Variable')
```

Comment:
I think it makes sense to allow one of them to be a number, just forbid both. E.g. you want your model to generate means and you want to sample with a fixed variance